File: gherkin/python/gherkin/token_scanner.py
import io
import os
import sys
from .token import Token
from .gherkin_line import GherkinLine


class TokenScanner(object):
    """
    The scanner reads a gherkin doc (typically read from a `.feature` file) and creates a token for
    each line.

    The tokens are passed to the parser, which outputs an AST (Abstract Syntax Tree).

    If the scanner sees a `#` language header, it will reconfigure itself dynamically to look for
    Gherkin keywords for the associated language. The keywords are defined in
    :file:`gherkin-languages.json`.
    """

    def __init__(self, path_or_str):
        if os.path.exists(path_or_str):
            self.io = io.open(path_or_str, 'rU', encoding='utf8')
        else:
            if sys.version_info < (3, 0):
                if isinstance(path_or_str, str):
                    self.io = io.StringIO(unicode(path_or_str, encoding='utf8'))
                else:
                    self.io = io.StringIO(path_or_str)
            else:
                self.io = io.StringIO(path_or_str)
        self.line_number = 0

    def read(self):
        self.line_number += 1
        location = {'line': self.line_number, 'column': 0}
        line = self.io.readline()
        return Token((GherkinLine(line, self.line_number) if line else line), location)

    def __del__(self):
        # close file descriptor if it's still open
        try:
            self.io.close()
        except AttributeError:
            pass

File: gherkin/javascript/src/legacy/gherkin/token_scanner.ts
import Token from './token'
import GherkinLine from './gherkin_line'
import { messages } from 'cucumber-messages'

/**
 * The scanner reads a gherkin doc (typically read from a .feature file) and creates a token for each line.
 * The tokens are passed to the parser, which outputs an AST (Abstract Syntax Tree).
 *
 * If the scanner sees a `#` language header, it will reconfigure itself dynamically to look for
 * Gherkin keywords for the associated language. The keywords are defined in gherkin-languages.json.
 */
export default class TokenScanner {
  private lineNumber = 0
  private lines: string[]

  constructor(source: string) {
    this.lines = source.split(/\r?\n/)
    if (
      this.lines.length > 0 &&
      this.lines[this.lines.length - 1].trim() === ''
    ) {
      this.lines.pop()
    }
  }

  public read() {
    const line = this.lines[this.lineNumber++]
    const location = messages.Location.fromObject({
      line: this.lineNumber,
      column: 0,
    })
    return line == null
      ? new Token(null, location)
      : new Token(new GherkinLine(line, this.lineNumber), location)
  }
}
